File: libs/go/src/domain/money.go
package domain

import (
	"encoding/json"
	"fmt"
	"math/big"
	"strings"
)

// Currency represents an ISO 4217 currency code.
type Currency string

// Common currency codes.
const (
	USD Currency = "USD"
	EUR Currency = "EUR"
	GBP Currency = "GBP"
	JPY Currency = "JPY"
	BRL Currency = "BRL"
	CNY Currency = "CNY"
	INR Currency = "INR"
	AUD Currency = "AUD"
	CAD Currency = "CAD"
	CHF Currency = "CHF"
)

// currencyDecimals maps currencies to their decimal places.
var currencyDecimals = map[Currency]int{
	USD: 2, EUR: 2, GBP: 2, JPY: 0, BRL: 2,
	CNY: 2, INR: 2, AUD: 2, CAD: 2, CHF: 2,
}

// Money represents a monetary value with currency.
type Money struct {
	amount   *big.Int // Amount in smallest unit (cents, etc.)
	currency Currency
}

// NewMoney creates a new Money from amount in smallest unit.
func NewMoney(amount int64, currency Currency) (Money, error) {
	if _, ok := currencyDecimals[currency]; !ok {
		return Money{}, fmt.Errorf("unsupported currency: %s", currency)
	}
	return Money{
		amount:   big.NewInt(amount),
		currency: currency,
	}, nil
}

// MustNewMoney creates a new Money, panicking on invalid input.
func MustNewMoney(amount int64, currency Currency) Money {
	m, err := NewMoney(amount, currency)
	if err != nil {
		panic(err)
	}
	return m
}

// NewMoneyFromFloat creates Money from a float value.
func NewMoneyFromFloat(amount float64, currency Currency) (Money, error) {
	decimals, ok := currencyDecimals[currency]
	if !ok {
		return Money{}, fmt.Errorf("unsupported currency: %s", currency)
	}
	multiplier := int64(1)
	for i := 0; i < decimals; i++ {
		multiplier *= 10
	}
	cents := int64(amount * float64(multiplier))
	return Money{
		amount:   big.NewInt(cents),
		currency: currency,
	}, nil
}

// Zero returns zero money for a currency.
func Zero(currency Currency) Money {
	return Money{amount: big.NewInt(0), currency: currency}
}

// Amount returns the amount in smallest unit.
func (m Money) Amount() int64 {
	if m.amount == nil {
		return 0
	}
	return m.amount.Int64()
}

// Currency returns the currency.
func (m Money) Currency() Currency {
	return m.currency
}

// IsZero returns true if the amount is zero.
func (m Money) IsZero() bool {
	return m.amount == nil || m.amount.Sign() == 0
}

// IsPositive returns true if the amount is positive.
func (m Money) IsPositive() bool {
	return m.amount != nil && m.amount.Sign() > 0
}

// IsNegative returns true if the amount is negative.
func (m Money) IsNegative() bool {
	return m.amount != nil && m.amount.Sign() < 0
}

// Add adds two Money values.
func (m Money) Add(other Money) (Money, error) {
	if m.currency != other.currency {
		return Money{}, fmt.Errorf("currency mismatch: %s vs %s", m.currency, other.currency)
	}
	result := new(big.Int).Add(m.amount, other.amount)
	return Money{amount: result, currency: m.currency}, nil
}

// Subtract subtracts another Money value.
func (m Money) Subtract(other Money) (Money, error) {
	if m.currency != other.currency {
		return Money{}, fmt.Errorf("currency mismatch: %s vs %s", m.currency, other.currency)
	}
	result := new(big.Int).Sub(m.amount, other.amount)
	return Money{amount: result, currency: m.currency}, nil
}

// Multiply multiplies by a scalar.
func (m Money) Multiply(factor int64) Money {
	result := new(big.Int).Mul(m.amount, big.NewInt(factor))
	return Money{amount: result, currency: m.currency}
}

// Divide divides by a scalar with rounding.
func (m Money) Divide(divisor int64) Money {
	result := new(big.Int).Div(m.amount, big.NewInt(divisor))
	return Money{amount: result, currency: m.currency}
}

// Equals checks if two Money values are equal.
func (m Money) Equals(other Money) bool {
	return m.currency == other.currency && m.amount.Cmp(other.amount) == 0
}

// Compare compares two Money values.
func (m Money) Compare(other Money) int {
	if m.currency != other.currency {
		return strings.Compare(string(m.currency), string(other.currency))
	}
	return m.amount.Cmp(other.amount)
}

// String returns a human-readable representation.
func (m Money) String() string {
	decimals := currencyDecimals[m.currency]
	if decimals == 0 {
		return fmt.Sprintf("%s %d", m.currency, m.amount.Int64())
	}
	divisor := int64(1)
	for i := 0; i < decimals; i++ {
		divisor *= 10
	}
	whole := m.amount.Int64() / divisor
	frac := m.amount.Int64() % divisor
	if frac < 0 {
		frac = -frac
	}
	return fmt.Sprintf("%s %d.%0*d", m.currency, whole, decimals, frac)
}

// moneyJSON is the JSON representation of Money.
type moneyJSON struct {
	Amount   int64    `json:"amount"`
	Currency Currency `json:"currency"`
}

// MarshalJSON implements json.Marshaler.
func (m Money) MarshalJSON() ([]byte, error) {
	return json.Marshal(moneyJSON{
		Amount:   m.Amount(),
		Currency: m.currency,
	})
}

// UnmarshalJSON implements json.Unmarshaler.
func (m *Money) UnmarshalJSON(data []byte) error {
	var j moneyJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return err
	}
	money, err := NewMoney(j.Amount, j.Currency)
	if err != nil {
		return err
	}
	*m = money
	return nil
}

File: libs/elixir/apps/auth_platform/lib/auth_platform/domain/money.ex
defmodule AuthPlatform.Domain.Money do
  @moduledoc """
  Monetary value with currency handling.

  Money values are stored as integers representing the smallest currency unit
  (e.g., cents for USD, pence for GBP).

  ## Usage

      alias AuthPlatform.Domain.Money

      # Creating money
      {:ok, price} = Money.new(1000, :USD)  # $10.00
      {:ok, tax} = Money.new(80, :USD)      # $0.80

      # Arithmetic
      {:ok, total} = Money.add(price, tax)  # $10.80

      # Formatting
      Money.format(total)  # "$10.80"

  ## Supported Currencies

  - USD (US Dollar)
  - EUR (Euro)
  - GBP (British Pound)
  - BRL (Brazilian Real)
  - JPY (Japanese Yen)

  """

  @type currency :: :USD | :EUR | :GBP | :BRL | :JPY

  @type t :: %__MODULE__{
          amount: integer(),
          currency: currency()
        }

  @enforce_keys [:amount, :currency]
  defstruct [:amount, :currency]

  @currencies [:USD, :EUR, :GBP, :BRL, :JPY]

  @currency_symbols %{
    USD: "$",
    EUR: "€",
    GBP: "£",
    BRL: "R$",
    JPY: "¥"
  }

  @currency_decimals %{
    USD: 2,
    EUR: 2,
    GBP: 2,
    BRL: 2,
    JPY: 0
  }

  @doc """
  Returns the list of supported currencies.
  """
  @spec supported_currencies() :: [currency()]
  def supported_currencies, do: @currencies

  @doc """
  Creates a new Money value.

  Amount should be in the smallest currency unit (cents, pence, etc.).

  ## Examples

      iex> AuthPlatform.Domain.Money.new(1000, :USD)
      {:ok, %AuthPlatform.Domain.Money{amount: 1000, currency: :USD}}

      iex> AuthPlatform.Domain.Money.new(1000, :INVALID)
      {:error, "unsupported currency: INVALID"}

      iex> AuthPlatform.Domain.Money.new("100", :USD)
      {:error, "amount must be an integer"}

  """
  @spec new(integer(), currency()) :: {:ok, t()} | {:error, String.t()}
  def new(amount, currency) when is_integer(amount) and currency in @currencies do
    {:ok, %__MODULE__{amount: amount, currency: currency}}
  end

  def new(_amount, currency) when currency not in @currencies do
    {:error, "unsupported currency: #{inspect(currency)}"}
  end

  def new(_amount, _currency) do
    {:error, "amount must be an integer"}
  end

  @doc """
  Creates a new Money value, raising on invalid input.
  """
  @spec new!(integer(), currency()) :: t()
  def new!(amount, currency) do
    case new(amount, currency) do
      {:ok, money} -> money
      {:error, reason} -> raise ArgumentError, reason
    end
  end

  @doc """
  Creates Money from a decimal amount (e.g., 10.50 for $10.50).

  ## Examples

      iex> AuthPlatform.Domain.Money.from_decimal(10.50, :USD)
      {:ok, %AuthPlatform.Domain.Money{amount: 1050, currency: :USD}}

  """
  @spec from_decimal(number(), currency()) :: {:ok, t()} | {:error, String.t()}
  def from_decimal(decimal, currency) when is_number(decimal) and currency in @currencies do
    decimals = Map.get(@currency_decimals, currency, 2)
    amount = round(decimal * :math.pow(10, decimals))
    new(amount, currency)
  end

  def from_decimal(_, currency) when currency not in @currencies do
    {:error, "unsupported currency: #{inspect(currency)}"}
  end

  @doc """
  Adds two Money values of the same currency.

  ## Examples

      iex> {:ok, a} = AuthPlatform.Domain.Money.new(100, :USD)
      iex> {:ok, b} = AuthPlatform.Domain.Money.new(50, :USD)
      iex> AuthPlatform.Domain.Money.add(a, b)
      {:ok, %AuthPlatform.Domain.Money{amount: 150, currency: :USD}}

      iex> {:ok, usd} = AuthPlatform.Domain.Money.new(100, :USD)
      iex> {:ok, eur} = AuthPlatform.Domain.Money.new(100, :EUR)
      iex> AuthPlatform.Domain.Money.add(usd, eur)
      {:error, "currency mismatch: USD vs EUR"}

  """
  @spec add(t(), t()) :: {:ok, t()} | {:error, String.t()}
  def add(%__MODULE__{currency: c, amount: a1}, %__MODULE__{currency: c, amount: a2}) do
    {:ok, %__MODULE__{amount: a1 + a2, currency: c}}
  end

  def add(%__MODULE__{currency: c1}, %__MODULE__{currency: c2}) do
    {:error, "currency mismatch: #{c1} vs #{c2}"}
  end

  @doc """
  Subtracts one Money value from another.

  ## Examples

      iex> {:ok, a} = AuthPlatform.Domain.Money.new(100, :USD)
      iex> {:ok, b} = AuthPlatform.Domain.Money.new(30, :USD)
      iex> AuthPlatform.Domain.Money.subtract(a, b)
      {:ok, %AuthPlatform.Domain.Money{amount: 70, currency: :USD}}

  """
  @spec subtract(t(), t()) :: {:ok, t()} | {:error, String.t()}
  def subtract(%__MODULE__{currency: c, amount: a1}, %__MODULE__{currency: c, amount: a2}) do
    {:ok, %__MODULE__{amount: a1 - a2, currency: c}}
  end

  def subtract(%__MODULE__{currency: c1}, %__MODULE__{currency: c2}) do
    {:error, "currency mismatch: #{c1} vs #{c2}"}
  end

  @doc """
  Multiplies a Money value by a factor.

  ## Examples

      iex> {:ok, price} = AuthPlatform.Domain.Money.new(100, :USD)
      iex> AuthPlatform.Domain.Money.multiply(price, 3)
      {:ok, %AuthPlatform.Domain.Money{amount: 300, currency: :USD}}

  """
  @spec multiply(t(), number()) :: {:ok, t()}
  def multiply(%__MODULE__{amount: amount, currency: currency}, factor) when is_number(factor) do
    {:ok, %__MODULE__{amount: round(amount * factor), currency: currency}}
  end

  @doc """
  Returns the decimal representation of the amount.

  ## Examples

      iex> {:ok, money} = AuthPlatform.Domain.Money.new(1050, :USD)
      iex> AuthPlatform.Domain.Money.to_decimal(money)
      10.5

  """
  @spec to_decimal(t()) :: float()
  def to_decimal(%__MODULE__{amount: amount, currency: currency}) do
    decimals = Map.get(@currency_decimals, currency, 2)
    amount / :math.pow(10, decimals)
  end

  @doc """
  Formats the Money value as a string with currency symbol.

  ## Examples

      iex> {:ok, money} = AuthPlatform.Domain.Money.new(1050, :USD)
      iex> AuthPlatform.Domain.Money.format(money)
      "$10.50"

      iex> {:ok, yen} = AuthPlatform.Domain.Money.new(1000, :JPY)
      iex> AuthPlatform.Domain.Money.format(yen)
      "¥1000"

  """
  @spec format(t()) :: String.t()
  def format(%__MODULE__{amount: amount, currency: currency}) do
    symbol = Map.get(@currency_symbols, currency, "")
    decimals = Map.get(@currency_decimals, currency, 2)

    if decimals == 0 do
      "#{symbol}#{amount}"
    else
      divisor = :math.pow(10, decimals)
      formatted = :erlang.float_to_binary(amount / divisor, decimals: decimals)
      "#{symbol}#{formatted}"
    end
  end

  @doc """
  Checks if the amount is zero.
  """
  @spec zero?(t()) :: boolean()
  def zero?(%__MODULE__{amount: 0}), do: true
  def zero?(%__MODULE__{}), do: false

  @doc """
  Checks if the amount is positive.
  """
  @spec positive?(t()) :: boolean()
  def positive?(%__MODULE__{amount: amount}), do: amount > 0

  @doc """
  Checks if the amount is negative.
  """
  @spec negative?(t()) :: boolean()
  def negative?(%__MODULE__{amount: amount}), do: amount < 0

  defimpl Jason.Encoder do
    def encode(%{amount: amount, currency: currency}, opts) do
      Jason.Encode.map(%{amount: amount, currency: currency}, opts)
    end
  end
end
